### PR TITLE
feat: add wire tagging

### LINKED ISSUE #236: [Feature] Access circuit wires and gates through tags
### DRI (owner)

@4l0n50 

### Flags

- [ ] perf
- [ ] needs-rfc

### RFC link (required if needs-rfc is checked)

_No response_

### Motivation / User value

- stakeholders: Developers
- value:  makes easier and cleaner testing/debugging.
- success: no `races.witness_trace.values[...]` in the code.

### Scope

in: circuit builder and tests
out: everything else

### Proposed approach

In tests and debugging, accessing the value of a wire currently requires reading
memory directly, for example `traces.witness_trace.values[i]`.

```rust
#[test]
fn field_add_identity(a in field_element()) {
    let mut builder = CircuitBuilder::<BabyBear>::new();
    let ca = builder.add_const(a);
    let zero = builder.add_const(BabyBear::ZERO);
    let result = builder.add(ca, zero);

    let circuit = builder.build().unwrap();
    let runner = circuit.runner();
    let traces = runner.run().unwrap();

    prop_assert_eq!(
        traces.witness_trace.values[result.0 as usize],
        a,
        "a + 0 = a"
    );
}
```
It would be more convenient to tag wires during circuit construction and retrieve
their values by name after execution.

``` rust
        #[test]
        fn field_add_identity(a in field_element()) {
            let mut builder = CircuitBuilder::<BabyBear>::new();
            let ca = builder.add_const(a);
            let zero = builder.add_const(BabyBear::ZERO);
            let result = builder.add(ca, zero);
            builder.probe(result, "Wire a")

            let circuit = builder.build().unwrap();
            let  runner = circuit.runner();
            let traces = runner.run().unwrap();

            prop_assert_eq!(
                runner.probe("Wire a"),
                a,
                "a + 0 = a"
            );
        }
```

A similar issue exists for non-primitive operations. Currently, `set_non_primitive_op_private_data` identifies operations using a `NonPrimitiveOpId`. Since private data is provided only after the circuit is built and run, the builder must return these IDs and the caller must process all of them in the correct order.
Using tags instead would be more natural. For example, when recursively proving a FRI proof, private data could be set using a tag like `"fri-query-55-depth-4-extra"`to identify the MMCS verification of an extra matrix
for query i at depth j.

### Performance / benchmarking (if relevant)

_No response_

### Exit criteria

- No more direct accesses to memory `races.witness_trace.values` in the code.
- No new tests. Tests using direct memory access should use probe instead.
- The doc for probe should be enough?

### Dependencies / Blockers

_No response_

### Notes / Links

_No response_

### PR DESCRIPTION
**Pull request by Nashtare**
## Summary
What does this PR change?

Adds a tagging mechanism in the `CircuitBuilder` to mark expressions, to be later on probed on the `Trace` side, for easier trace cell access and for private data access by NPOs.

Fixes #236.

## Why

Solves complicated trace accesses during debugging and simplifies future NPOs private-data setting.

## Changes
- [x] Add a `tag` method to `CircuitBuilder` with a user-provided string (can be dynamic). Note that this is *different* from the allocation tags used alternatively for debugging.
- [x] Add a `probe` method on `Trace` to get the value of a tagged wire.
- [x] API change: reduced `set_non_primitive_op_private_data` to `set_private_data` as the NPO part is implied, and allows to keep the new method `set_private_data_by_tag` short.

## Tests
- [x] Unit tests added/updated
- [x] Existing tests pass
- [x] Edge cases considered: checked duplicated tagging, and disappearing tag post circuit optimization

## Benchmarks / Performance (if relevant)
- [ ] Before/after numbers included
- [ ] Benchmark command or link to CI perf output
- [x] No unexpected regressions in CPU / memory / proof size / compile time: no regression detected on simple tests, but there may be a slight overhead for tagging-heavy NPOs in the future. TBD

## Docs (if relevant)
- [x] Rustdoc updated
- [ ] Book/README updated
- [x] Examples updated

### CODE REVIEW

File: circuit/src/builder/circuit_builder.rs [hunk 1]
```diff
@@ -564,6 +624,20 @@ where
         circuit.enabled_ops = self.config.into_enabled_ops();
         circuit.non_primitive_trace_generators = self.non_primitive_trace_generators;
 
+        // Transfer wire tags, converting ExprId to WitnessId
+        for (tag, expr_id) in self.tag_to_expr {
+            if let Some(&witness_id) = circuit.expr_to_widx.get(&expr_id) {
+                circuit.tag_to_witness.insert(tag, witness_id);
+            } else {
+                tracing::warn!(
+                    "Tag {tag} dropped after circuit optimization: expr_id {expr_id} not found"
```
[the remark this answers is not in the document]
Comment (Nashtare) [reply circuit/src/builder/circuit_builder.rs]: I don't see why it should return an error. Eventually instead of just warning that the targeted `ExprId` was deleted after the optimization pass we could rebind the tag to the matching one as you suggested

[the remark this answers is not in the document]
Comment (4l0n50) [reply circuit/src/builder/circuit_builder.rs]: ah yes sorry, when I added the second sentence I forgot to remove the first. so ok if no error as long as we find the tag's expr_id

File: circuit/src/builder/circuit_builder.rs [hunk 2]
```diff
@@ -521,6 +529,58 @@ where
     pub fn list_scopes(&self) -> Vec<&'static str> {
         self.expr_builder.list_scopes()
     }
+
+    /// Tags an expression for value lookup via `Traces::probe()` later on during
+    /// circuit execution.
+    ///
+    /// Tags must be unique within a circuit. Duplicate tags will return an error.
+    ///
+    /// Note that this is different from allocation labels for `ExprId`s, which are
```
[the remark this answers is not in the document]
Comment (4l0n50) [reply circuit/src/builder/circuit_builder.rs]: I'm ok with it then. thanks!
